### PR TITLE
SSL Pretrain v2 featurization 

### PR DESCRIPTION
**Pull request by jiali1025**
## Description
The core updates include the new pretrainmolgraph object and pretrain batch collate methods, which do not affect the existing functionality. These updates have been thoroughly tested. As suggested, the SSL_v2 function will be implemented through multiple PRs to facilitate easier review, as it involves adding a significant amount of code in stages.

### CODE REVIEW

**Comment by KnathanM** [comment]
Hi @jiali1025 could you include a description of some of the differences between the new pretrainmolgraph object and pretrain batch collate methods and the versions currently in Chemprop, along with an explanation of why those differences are required?

**Comment by jiali1025** [comment]
> Hi @jiali1025 could you include a description of some of the differences between the new pretrainmolgraph object and pretrain batch collate methods and the versions currently in Chemprop, along with an explanation of why those differences are required?

Hi Nathan, no problem I will do tmr, it is late my time. Could u let me know whether I just comment here or comment inside the code?

**Comment by jiali1025** [comment]
> Hi @jiali1025 could you include a description of some of the differences between the new pretrainmolgraph object and pretrain batch collate methods and the versions currently in Chemprop, along with an explanation of why those differences are required?

@KnathanM The primary differences involve the inclusion of customized methods for atom masking, bond deletion, and subgraph deletion. These operations modify the original graph by either masking features or altering connections. In the self-supervised learning approach we developed for v1SSL, masked atom prediction is a key task. Therefore, it is necessary to dynamically record the corresponding labels and masked indices, as these operations are performed randomly. This kind of information is not included in the original v2 molgraph, batch, or featurization. Bond deletion and subgraph deletion are mainly used in the contrastive learning part of the self-supervised learning process. I believe Jackson and Kevin are familiar with the details. Using SSL pre-training will enhance the model’s performance across multiple tasks.

**Comment by davidegraff** [comment]
An open question about self-supervision implementation. My understanding here is that you're implementing via a custom data structure that handles the masking logic, but this is counter to v2 object model that simple data structures generally don't possess instance methods which mutate the underlying data. Just thinking high-level here, but is it possible to implement self-supervision masking via a suite of `nn.Module`s? For example:

```python
class NodeMasking(nn.Module):
    def forward(self, G: Graph) -> Graph:
        ...

class EdgeMasking(nn.Module):
    def forward(self, G: Graph) -> Graph:
        ...

class SubgraphMasking(nn.Module):
    def forward(self, G: Graph) -> Graph:
        ...
```

**Review by KnathanM — COMMENTED** [review]
Instead of masking the graphs before collating them, can we mask them after? If it is possible to mask them after then we wouldn't need to change any of `collate.py`, `molgraph.py`, and `molecule.py` which would be simpler. 

Focusing first on node masking, I see that your `PreBatchMolGraph` has the same attributes as `BatchMolGraph` plus a `batch_masked_atom_index` attribute and a `batch_masked_atom_label_list` attribute. We can generate these two extra attributes from the `bmg: BatchMolGraph` after collating by using a `torch.nn.Module` if we assume we can get the atom identity from `bmg.V` based on knowledge of the atom featurizer. I admit that is a bit error prone, but I'd rather think of ways to make that less error prone than add several new large classes if we can avoid it. 

Here is the `torch.nn.Module` I came up with for node masking:
```
from torch import nn, Tensor
import torch
from chemprop.data import BatchMolGraph

class NodeMasking(nn.Module):
    def __init__(self, mask_atom_pre_percent: float):
        super().__init__()
        self.mask_atom_pre_percent = mask_atom_pre_percent

    def forward(self, bmg: BatchMolGraph) -> tuple[BatchMolGraph, Tensor, Tensor]:
        n_atoms_per_graph = torch.bincount(bmg.batch)
        offset = torch.cumsum(n_atoms_per_graph, dim=0)
        offset[-1] = 0

        masked_atom_indices = [
            torch.randperm(n)[:int(n * self.mask_atom_pre_percent)] for n in n_atoms_per_graph
            ]
        masked_atom_indices = [
            indices + offset[i-1] for i, indices in enumerate(masked_atom_indices)
            ]
        batch_masked_atom_index = torch.cat(masked_atom_indices)
        batch_masked_atom_label_list = bmg.V[batch_masked_atom_index, :38].argmax(dim=1)
        bmg.V[batch_masked_atom_index] = 0.0

        return bmg, batch_masked_atom_index, batch_masked_atom_label_list
```
The `batch_masked_atom_label_list` contains the indices of 1 in the one-hot atom type encoding. This isn't the same as the atomic number, but the atom type prediction is just to pretrain anyways. It is hardcoded to look in the first 38 columns of the `bmg.V` array because that is the length of the default atom featurizers atom type one-hot encoding. It would be better if this value could be dynamically gotten based on knowledge of the featurizer used. I assume all atom featurizers will include the atom type as a feature...

An instance of this class could be used before the message passing layer:
```
masknodes = NodeMasking(mask_atom_pre_percent=0.3)
model = SSLMPNN(..., node_masker=masknodes)
```
```
class SSLMPNN():
    def training_step(self, batch):
        bmg, V_d, X_d, targets, weights, lt_mask, gt_mask = batch
        bmg, batch_masked_atom_index, batch_masked_atom_label_list = self.node_masker(bmg)
        H_v = self.message_passing(bmg, V_d)
        ...
```

There might be issues with bond or subgraph masking that I am missing as I haven't looked at that. What do you think @jiali1025?

**Comment by JacksonBurns** [comment]
Making a note from some offline discussion - with the addition of the `--from-foundation` argument, we are considering taking a trained foundational model that uses SSL and making it available in v2.2.n without actually merging in the code to train _more_ SSL models.
